cmd/juju/user/change_password.go

```diff
@@ -92,78 +120,129 @@ func (c *changePasswordCommand) Run(ctx *cmd.Context) error {
 		defer c.api.Close()
 	}
 
-	newPassword, err := readAndConfirmPassword(ctx)
-	if err != nil {
-		return errors.Trace(err)
+	if c.Reset {
+		return c.resetUserPassword(ctx)
 	}
+	return c.updateUserPassword(ctx)
+}
 
-	controllerName, err := c.ControllerName()
+func (c *changePasswordCommand) prepareRun() error {
+	err := c.ensureControllerName()
 	if err != nil {
 		return errors.Trace(err)
 	}
-	store := c.ClientStore()
-	accountDetails, err := store.AccountDetails(controllerName)
+
+	c.accountDetails, err = c.ClientStore().AccountDetails(c.controllerName)
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	var userTag names.UserTag
 	if c.User != "" {
 		if !names.IsValidUserName(c.User) {
 			return errors.NotValidf("user name %q", c.User)
 		}
-		userTag = names.NewUserTag(c.User)
-		if userTag.Id() != accountDetails.User {
+		c.userTag = names.NewUserTag(c.User)
+		if c.userTag.Id() != c.accountDetails.User {
 			// The account details don't correspond to the username
 			// being changed, so we don't need to update the account
 			// locally.
-			accountDetails = nil
+			c.accountDetails = nil
 		}
 	} else {
-		if !names.IsValidUser(accountDetails.User) {
-			return errors.Errorf("invalid user in account %q", accountDetails.User)
+		if !names.IsValidUser(c.accountDetails.User) {
+			return errors.Errorf("invalid user in account %q", c.accountDetails.User)
 		}
-		userTag = names.NewUserTag(accountDetails.User)
-		if !userTag.IsLocal() {
-			return errors.Errorf("cannot change password for external user %q", userTag)
+		c.userTag = names.NewUserTag(c.accountDetails.User)
+		if !c.userTag.IsLocal() {
+			operation := "change"
+			if c.Reset {
+				operation = "reset"
+			}
+			return errors.Errorf("cannot %v password for external user %q", operation, c.userTag)
 		}
 	}
-	if err := c.api.SetPassword(userTag.Id(), newPassword); err != nil {
+	return nil
+}
+
+func (c *changePasswordCommand) ensureControllerName() error {
+	controllerName, err := c.ControllerName()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	c.controllerName = controllerName
+	return nil
+}
+
+func (c *changePasswordCommand) resetUserPassword(ctx *cmd.Context) error {
+	key, err := c.api.ResetPassword(c.userTag.Id())
+	if err != nil {
 		return block.ProcessBlockedError(err, block.BlockChange)
 	}
+	var macaroonErr error
+	registerMsg := "Ask the user to run:"
+	if c.accountDetails == nil {
+		ctx.Infof("Password for %q has been reset.", c.User)
+	} else {
+		registerMsg = "Please run:"
+		if c.accountDetails.Password != "" {
+			macaroonErr = c.ClearControllerMacaroons(c.ClientStore(), c.controllerName)
+			if macaroonErr != nil {
+				macaroonErr = errors.Annotatef(err, "could not clear macaroon")
```

Comment:
I do wonder if we want to expose the user to the concept of macaroon. Or if we should just say:
"could not clear local credential cache".
Not a big deal either way, just something to consider.